spoc/query_engine.py

```diff
@@ -172,83 +199,367 @@ def __call__(self, genomic_data: GenomicData) -> GenomicData:
         regions = self._convert_to_duckdb(self._regions)
         # get position columns and construct filter
         position_fields = input_schema.get_position_fields()
+        # add end position if not present
+        if len(position_fields[1]) == 2:
+            genomic_df = self._add_end_position(
+                genomic_df, input_schema.get_binsize(), position_fields
+            )
+            position_fields = {
+                position: [f"chrom_{position}", f"start_{position}", f"end_{position}"]
+                for position in position_fields.keys()
+            }
         # construct query
-        df = genomic_df.set_alias("data").join(
+        snipped_df = genomic_df.set_alias("data").join(
             regions.set_alias("regions"), self._contstruct_filter(position_fields)
         )
-        return QueryResult(df, self._get_transformed_schema(input_schema))
+        return QueryPlan(
+            snipped_df,
+            self._get_transformed_schema(snipped_df, input_schema, position_fields),
+        )
+
 
+class AggregationFunction(Enum):
```

Comment:
Wouldn't it be too wordy to specify a non-default agg function in `OffsetAggregtaion`?

Reply:
I would implement that as a separate type that can be passed as a function, for example, a basic implementation would be that we allow `Union[AggregationFunction|str]` for the function argument and if it's a string, we call `.format` with the value column. So if you want something like `MIN(columns)`, you would pass `MIN({})`.

Reply:
Okay, that sounds reasonable for custom functions. I was a bit confused that if I want to specify a non-default function (like SUM), I will have to type `AggregationFunction.SUM`, which is too long. But then we can come up with an alias, like `import torch.nn.functional as F`.

For the case that you described, passing smth like `MIN({})` seems to be a bit redundant -- we can attach `({})` under the hood.

Speaking of custom functions, we have two sorts of them: defined in duckdb and UDFs. Former should be simple to use, while I am curious how UDFs are defined and used in duckdb (since one can write a duckdb UDF in python, this seems to be a nice feature for the future).

Reply:
Inded, UDFs would be nice in the future, but would leave that for now